src/app/guards/userRole.guard.ts

```diff
@@ -0,0 +1,38 @@
+import { CanActivateFn, Router } from "@angular/router";
+import { Option } from "../shared/option.model";
+import { AuthService } from "../auth/auth.service";
+import { inject } from "@angular/core";
+import { User } from "../users/user.model";
+
+export const userRoleGuard = (criteriaRoles: Option[], whitelist: boolean): CanActivateFn => {
+  
+  return async (route, state): Promise<boolean> => {
+  
+    const router = inject(Router);
+
+    // Check if user is authenticated
+    const authenticatedUser: User | null = inject(AuthService).currentUserSubject.getValue();  
+    if(!authenticatedUser) { 
+      router.navigate(['/login']);
+      throw Error("User is null in provided AuthService"); 
+    }
+
+    // Check if user has appropriate role
+    if(whitelist) {
+      for (let i = 0; i < criteriaRoles.length; i++) {
```

Comment:
you can use js array functions like forEach and includes to minimise the amount of code